src/CostFunction3D.h

```diff
@@ -151,9 +151,16 @@ CostFunction3D(const Projection& proj, const int& numObs = 0, const int& stateSi
 	double *iFFTin, *jFFTin, *kFFTin;
 	fftw_complex *iFFTout, *jFFTout, *kFFTout;
         bool UseFFT;
+
+	// explicitly store the H matrix in CSR format
 	real *H;
-	integer *IH, *I2H,*JH;
-  integer *mPtr, *mVal;
+	uint64_t *IH; // uint64_t
+        uint32_t *JH; // uint32_t
```

Comment:
why shall we define `*IH` and `*JH` with different types?

Reply:
Good question and this relates to your overall question as to why these changes save memory.  The JH array can have values that range from (0,nstate-1) while IH can have values that range from (0,nnz-1).  While nnz exceeds the threshold for 32-bit integers, nState does not.  So we can save memory here by using 32-bit inttegers for the JH array which has a total number of elements of nnz.  The previous implementation had multiple 64-bit integer index arrays that had nnz elements.  I have added comments to the CostFunction3D.h to help clarify the differences in word size.

Reply:
Aha, I see. Thanks for your clarification and it now makes sense. As long as `nState` won't exceed the threshold for 32-bit integer in the future SAMURAI production test, this change seems safe to me.